beacon-chain/core/validators/validator.go

```diff
@@ -403,15 +383,54 @@ func ProcessPenaltiesAndExits(ctx context.Context, state *pb.BeaconState) *pb.Be
 			totalAtEnd := state.LatestSlashedBalances[slashedEpoch]
 			totalPenalties := totalAtStart - totalAtEnd
 
-			penaltyMultiplier := totalPenalties * 3
-			if totalBalance < penaltyMultiplier {
-				penaltyMultiplier = totalBalance
+			var penalty uint64
+			index := uint64(idx)
+			if totalPenalties*3 < totalBalance {
+				if helpers.EffectiveBalance(state, index)*totalPenalties*3/totalBalance > helpers.EffectiveBalance(state, index)/params.BeaconConfig().MinPenaltyQuotient {
+					penalty = helpers.EffectiveBalance(state, index) * totalPenalties * 3 / totalBalance
+				} else {
+					penalty = helpers.EffectiveBalance(state, index) / params.BeaconConfig().MinPenaltyQuotient
+				}
+			} else {
+				if helpers.EffectiveBalance(state, index)*totalBalance/totalBalance > helpers.EffectiveBalance(state, index)/params.BeaconConfig().MinPenaltyQuotient {
+					penalty = helpers.EffectiveBalance(state, index) * totalBalance / totalBalance
+				} else {
+					penalty = helpers.EffectiveBalance(state, index) / params.BeaconConfig().MinPenaltyQuotient
+				}
```

Comment:
These lines are really long, maybe you can turn `helpers.EffectiveBalance(state, index)` into a local var, and re-use that? Same for `params.BeaconConfig().MinPenaltyQuotient`